### DISCUSSION TITLE
With WordPad being planned to be removed from Windows how do we generate rtf for EULA

### DISCUSSION THREAD

**Discussion by MikeWilliams-UK** [wixtoolset/issues#7705] [Questions]
At the weekend I saw this article [Microsoft is removing WordPad from Windows after nearly 30 years](https://www.theverge.com/2023/9/3/23857331/microsoft-wordpad-windows-removal-end-of-support) on the verge.

This is ringing alarm bells for our open source project [Chem4Word](https://github.com/Chem4Word)

We currently use WiX 3.11 in combination with Visual Studio 2022 and WiX toolset extension to generate our installer (Setup.msi)

When we first started to use WiX we quickly discovered that WiX 3.* produced a blank EULA screen with the rtf file which we generated using MS Word.

Google showed us that the fix was to use WordPad to write the rtf files as MS Word was putting in information which WiX did not understand.

**My Question is therefore**

What versions of WiX support rtf files written by MS Word for the EULA?
If none then this is a serious issue going forward and needs to be addressed.

**Comment by robmen** [comment]
I had the same thought when I saw that article. :)

Unfortunately, this isn't an issue in the WiX codebase. The issue is in the Windows Installer itself. Its RTF control doesn't like advanced RTF content.

I'm sure something new will come along to help clean up complex RTF files. Maybe someone will even help contribute such a clean-up mechanism to WiX so we can auto-fix complex RTF.

Or maybe Microsoft will fix the RTF control in the Windows Installer to handle more complex RTF. 🤣

**Reply by MikeWilliams-UK** [reply]
Does anyone have a definition of what kind of RTF is too complex for the installer?

**Reply by MikeWilliams-UK** [reply]
Maybe MS will make WordPad open source when it is removed, then it could be maintained for the purpose of generating EULA rtf documents.

**Reply by BMurri** [reply]
[This page](https://learn.microsoft.com/en-us/windows/win32/msi/scrollabletext-control) gives us a clue: `RICHEDIT class`.

[This page](https://learn.microsoft.com/en-us/windows/win32/controls/about-rich-edit-controls) tells us what versions of RTF the platform supports. We know it's not 4.1. Whether it is 1.0 or 3.0 depends on which DLL MSI loads, which can be determined with some experimentation (and tools from SystemInternals).

How to get Word to output a specific version of RTF, I have no clue.

**Reply by BMurri** [reply]
Based on what's in [this page](https://en.wikipedia.org/wiki/Rich_Text_Format), I'd say using `Word 6.0` is your safest bet (or, probably better, save a copy of `WordPad` for posterity).

I'm not sure running an Office product from 1993 would be wise, to say the least).

**Reply by MikeWilliams-UK** [reply]
I would err on the side of running WordPad in a Windows 7 VM to ensure it never got patched into oblivion.

**Reply by MikeWilliams-UK** [reply]
Running WordPad on my Windows 10 (22H2) laptop shows that it loads msftedit.dll from C:\Windwos\System32 folder.
It's properties say it is version 10.0.190441.1288

System Internals Process Explorer shows a lot of strings which I think look like the set of rtf groups which are supported.

This has me wondering if a simple C# program to keep the listed groups would do the trick ...

**Reply by MikeWilliams-UK** [reply]
This just dropped into my news feed
https://www.laptopmag.com/news/wordpad-is-dead-after-30-years-heres-3-great-alternatives

I had already thought about testing libre office, but Google docs may be a possibility...

**Reply by MikeWilliams-UK** [reply]
Just tried using Google Docs to produce an RTF.

I tested it with just one line of text as shown below
"This is an RTF produced by Google Docs"

WordPad - 1KB
Google Docs - 12KB
Word - 42KB

Not had a chance to test to see if the Google Docs version (which you need to "download as rtf") works

**Reply by MikeWilliams-UK** [reply]
Just downloaded and tried out LibreOffice Portable from https://www.libreoffice.org/download/portable-versions

With this a simple one line rtf weighs in at 3KB, and looks quite "simple" so might be the best alternative
Does seem to be a sledge hammer to crack a nut though

**Reply by BMurri** [reply]
Running `msiexec.exe` on my Windows 11 (fully updated, as of today) on an MSI with a license in a dialog shows that it (v`5.0.22621.1`) loaded version `5.31.23.1231` of `riched20.dll` which describes itself as `Rich Text Edit Control, v3.1`. Interestingly that DLL has a copyright statement that includes the year-span `1997-2008` (this for a file that appears to have been updated on my system by Microsoft from sources in May 2022).

So, referring to the [Rich Edit Version 3.0](https://learn.microsoft.com/en-us/windows/win32/controls/about-rich-edit-controls#rich-edit-version-30) feature chart (from before) you will want to find an editor that limits itself to that version of RTF (basically what Wikipedia describes as v1.3).

A WordPad in a VM running the latest released version of Windows XP, isolated from everything but its own host, may be the safest, most effective editor.

Alternately, if someone wants to use that same DLL and create a simple editor (because that is the "purpose" of that DLL), it could go viral around the Windows Installer community.

**Reply by MikeWilliams-UK** [reply]
Here's a radical thought - Why doesn't MS add a switch/option to SaveAs RTF to choose which RTF version to save the file as...
But failing that I think a VM running old version of windows is best bet.

**Reply by MikeWilliams-UK** [reply]
Sysinternals process explorer shows that on my Windows 10 22H2 work laptop running an installer (msiexec.exe version 5.0.19041.2193) has loaded the same loaded version 5.31.23.1231 of riched20.dll as @BMurri

**Reply by MikeWilliams-UK** [reply]
Just found a UWP app on the Windows Store called "TowPad"
https://apps.microsoft.com/store/detail/towpad/9MTMB5PDZV2M

I have just tried it out and modifying our EULA does not balloon it like Word does
Source code for it is avaliable on GitHub at https://github.com/itsWindows11/TowPad

It is not perfect - one example being that when you hover over the scroll bars the cursor remains as the text I-Beam.

**Reply by MikeWilliams-UK** [reply]
Just found another one which is a traditional Windows Forms App in GitHub at https://github.com/vovabatsyk/WordPad
This one looks more antiquated, but does the just the same.

**Comment by jcasale** [comment]
You could maintain your source document in Markdown format and use Pandoc to programmatically convert it to rich text format.

```batch
pandoc -s eula.md -o eula.rtf
```
See <https://pandoc.org/demos.html> for more info.
